text/0000-task-builder.md

```diff
@@ -0,0 +1,200 @@
+
+Any setup that requires the V8 VM (e.g., getting an `f64` from a `Number`) must be performed outside of this closure.
+
+The closure is expected to return _another closure_. This higher-order method will be familiar to many Javascript developers. The returned closure is the `complete` method and will be performed back on the main Javascript thread. In the `complete` closure, the user may convert Rust types back to Javascript types.
+
+Finally, a `schedule` method is provided to submit the task to the queue. For conveinance, it returns `JsUndefined`, matching Javascript semantics for most asynchronous methods.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+### Changes to low-level Task
+
+The existing low-level `Task` implementation in C++ is modified to match a new ownership model. The `perform` model will _consume_ self. The `complete` method must be changed to no longer pass `self` to prevent use-after-free.
+
+```rust
+pub fn Neon_Task_Schedule(task: *mut c_void,
+                          perform: unsafe extern fn(*mut c_void) -> *mut c_void,
+                          complete: unsafe extern fn(*mut c_void, &mut Local),
+                          callback: Local);
+```
+
+### Private `BaseTask` trait
+
+A new (private) `BaseTask` trait is introduced. The `BaseTask` trait is very similar to the existing `Task` trait except `fn perform` is changed to take ownership of `self` and `fn complete` is changed to an associated method (does not use self).
+
+A single `Output` parameter is provided and it is up to the user to return a `Result` if desired.
+
+```rust
+trait BaseTask: Send + Sized + 'static {
+    type Output: Send + 'static;
+    type JsEvent: Value;
+
+    fn perform(self) -> Self::Output;
+    fn complete<'a>(cx: TaskContext<'a>, result: Self::Output) -> JsResult<Self::JsEvent>;
+    fn schedule(self, callback: Handle<JsFunction>);
+}
+```
+
+### `Task` trait
+
+An implementation of `BaseTask` is added for `T: Task` to provide backwards compatibility to existing implementations of `Task`. Since `complete` is an associated method, `self` is threaded through the `Output` parameter.
+
+The existing `Task` trait will be _deprecated_.
+
+```rust
+impl<T: Task> BaseTask for T {
+    // Output contains a `Result` to match the `Task` trait
+    type Output = (Result<T::Output, T::Error>, T);
+    type JsEvent = T::JsEvent;
+
+    fn perform(self) -> Self::Output {
+        // `self` is threaded through `Output` as part of a tuple
+        (Task::perform(&self), self)
+    }
+
+    fn complete<'a>(cx: TaskContext<'a>, output: Self::Output) -> JsResult<Self::JsEvent> {
+        // `task` is unpacked from the tuple to use the `complete` method
+        let (result, task) = output;
+
+        task.complete(cx, result)
+    }
+}
+```
+
+### `TaskBuilder`
+
+The `TaskBuilder` struct contains two pieces of key data:
```

Comment:
Can you include the full `TaskBuilder` type declaration and the signatures of the methods in the RFC?